Initial services controller

MettyS opened the pull request.

MettyS committed:
Add services and central controller.

Initial domain types.

jer-irl commented on server/controller/central_checkers_controller.go:
```diff
@@ -0,0 +1,92 @@
+	"context"
+)
+
+// Tile int
```
Can you remove this comment?  And other comments that are redundant/unnecessary?

MettyS replied:
I updated the comments!

jer-irl commented on server/controller/central_checkers_controller.go:
```diff
@@ -0,0 +1,92 @@
+	PrevMoveSet []Move
+}
+
+// GameRole int
```
This comment seems unnecessary as well to me.

MettyS replied:
Updated

MettyS committed:
Initial Services

Initial services and type conversions

Registration of services in main.go

jer-irl commented on server/main.go:
```diff
@@ -3,31 +3,20 @@ package main
-	return res
-}
-
 // BoardUpdateSubscription
 
 // rpc MakeMoves(MoveRequest) returns (MoveResponse) {}
```
These comments weren't added before in this changeset, but they should be removed eventually

MettyS replied:
Removed!

jer-irl commented on server/controller/central_checkers_controller.go:
```diff
@@ -0,0 +1,92 @@
+// GameStartRole message
+type GameStartRole struct {
+	PlayerRole GameRole
+	GameID     string
```
Is offsetting the types like this a good style?  I don't personally like it if the linter didn't suggest/require it.

MettyS replied:
It is built into the golang linter and gofmt.

jer-irl commented on server/services/gamecontrol.go:
```diff
@@ -0,0 +1,37 @@
+	pb.UnimplementedGameControlServiceServer
+}
+
+func convertGameStartRequestInward(req *pb.GameStartRequest) *controller.GameStartRequest {
```
We should look up if there's a recommended max line width in go

jer-irl replied:
Actually, we should just use `go fmt` and use what it gives.

jer-irl commented on server/services/gamecontrol.go:
```diff
@@ -0,0 +1,37 @@
+	pb.UnimplementedGameControlServiceServer
+}
+
+func convertGameStartRequestInward(req *pb.GameStartRequest) *controller.GameStartRequest {
```
Why don't these functions return values directly, rather than refs/pointers?  It seems like values would be better if possible.

MettyS replied:
I refactored the services and controller to have internal-level avoid references. This is a good idea.

jer-irl commented on server/services/gameplay.go:
```diff
@@ -0,0 +1,51 @@
+}
+
+func convertMoveRequestInward(req *pb.MoveRequest) *controller.MoveRequest {
+	domMoveRequest := &controller.MoveRequest{}
```
As mentioned in another file, I think we should use values everywhere it's possible, so refs/pointers signal something else is going on.

MettyS replied:
Done.

jer-irl reviewed (changes requested):
Mostly just formatting and stylistic changes requested!

jer-irl commented on server/services/gameplay.go:
```diff
@@ -26,26 +26,61 @@ func convertMoveRequestInward(req *pb.MoveRequest) *controller.MoveRequest {
+	domMoveResponse, err := controller.HandleMakeMoves(ctx, domMoveRequest)
 
-	return convertMoveResponseOutward(domMoveResponse), nil
+	return convertMoveResponseOutward(domMoveResponse), err
```
I _think_ you're supposed to do an early return with `nil, err` if err is non-nil, but can you look that up?

MettyS replied:
Done!

jer-irl commented on server/services/gameplay.go:
```diff
@@ -26,26 +26,61 @@ func convertMoveRequestInward(req *pb.MoveRequest) *controller.MoveRequest {
+	pbBoardUpdate := pb.BoardUpdate{}
+
+	pbBoard := pb.BoardState{
+		Board: make([]pb.Tile, 32),
```
Magic number, can we define a constant out-of-line?

MettyS replied:
Defined in controller

jer-irl commented on server/services/gameplay.go:
```diff
@@ -26,26 +26,61 @@ func convertMoveRequestInward(req *pb.MoveRequest) *controller.MoveRequest {
+		pbMoves = append(pbMoves, &tempMove)
+	}
+	pbBoardUpdate.PrevMoveSet = pbMoves
+	return &pbBoardUpdate
```
Why return a pointer rather than a value?

jer-irl commented on server/services/gamecontrol.go:
```diff
@@ -12,13 +12,13 @@ type GameControlServer struct {
 }
 
-func convertGameStartResponseOutward(req *controller.GameStartResponse) *pb.GameStartResponse {
+func convertGameStartResponseOutward(req controller.GameStartResponse) *pb.GameStartResponse {
```
Does this have to return a pointer?

jer-irl commented on server/services/gameplay.go:
```diff
@@ -26,26 +26,61 @@ func convertMoveRequestInward(req *pb.MoveRequest) *controller.MoveRequest {
 }
 
-func convertMoveResponseOutward(res *controller.MoveResponse) *pb.MoveResponse {
+func convertMoveResponseOutward(res controller.MoveResponse) *pb.MoveResponse {
```
Does this have to return a pointer?

MettyS committed:
Update services to translate references.

Update comments to be more explanatory.
Fill out gameplay.BoardUpdateSubscription
implementation.

MettyS committed:
Error and magic number fix

Add early return on error

Create BoardSize constant
